File: src/genflow/models/database_adapter.py
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Tuple, Type


class DatabaseAdapter(ABC):
    @abstractmethod
    def create_table(
        self,
        table_name: str,
        key_schema: Dict[str, str],
        attribute_definitions: Dict[str, str],
        global_secondary_indexes: Dict[str, Dict[str, str]] | None = None,
        read_capacity_units: int = 1,
        write_capacity_units: int = 1,
    ) -> None:
        pass

    @abstractmethod
    def drop_table(self, table_name: str) -> None:
        pass

    @abstractmethod
    def get_primary_key(self) -> str:
        pass

    @abstractmethod
    def save(self, item: Dict[str, Any]) -> None:
        pass

    @abstractmethod
    def get(self, key: Any) -> Dict[str, Any] | None:
        pass

    @abstractmethod
    def delete(self, primary_key: Any) -> None:
        pass

    @abstractmethod
    def query(
        self,
        condition: str,
        values: Dict[str, Any],
        limit: int = 100,
        reverse: bool = False,
        start_key: str | None = None,
        index: str | None = None,
    ) -> tuple[list[dict[str, Any]], str]:
        pass

File: packages/models/src/database-adapter.ts
/**
 * Database Adapter – abstract interface for persistence backends.
 *
 * Port of Python's `nodetool.models.database_adapter`.
 */

import type { ConditionBuilder } from "./condition-builder.js";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Row = Record<string, any>;

export interface TableSchema {
  table_name: string;
  primary_key?: string; // defaults to "id"
   
  columns: Record<string, FieldDef>;
}

export interface FieldDef {
  type: "string" | "number" | "boolean" | "json" | "datetime";
  optional?: boolean;
  default_value?: unknown;
}

export interface IndexDef {
  name: string;
  columns: string[];
  unique: boolean;
}

/**
 * Abstract interface every database backend must implement.
 */
export interface DatabaseAdapter {
  readonly tableName: string;
  readonly tableSchema: TableSchema;

  /** Create the table (no-op if it already exists). */
  createTable(): Promise<void>;

  /** Drop the table. */
  dropTable(): Promise<void>;

  /** Insert or replace a row. */
  save(item: Row): Promise<void>;

  /** Retrieve a single row by primary key. */
  get(key: string | number): Promise<Row | null>;

  /** Delete a row by primary key. */
  delete(primaryKey: string | number): Promise<void>;

  /**
   * Query rows with optional filtering, ordering and pagination.
   *
   * @returns A tuple of [rows, cursor]. Cursor is "" when there are no more results.
   */
  query(opts?: {
    condition?: ConditionBuilder;
    orderBy?: string;
    limit?: number;
    reverse?: boolean;
    columns?: string[];
  }): Promise<[Row[], string]>;

  /** Create an index. */
  createIndex(
    indexName: string,
    columns: string[],
    unique?: boolean,
  ): Promise<void>;

  /** Drop an index. */
  dropIndex(indexName: string): Promise<void>;

  /** List all indexes on the table. */
  listIndexes(): Promise<IndexDef[]>;

  /** Get the primary key column name. */
  getPrimaryKey(): string;
}
